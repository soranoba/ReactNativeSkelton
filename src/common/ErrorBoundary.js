import React from 'react';
import PropTypes from 'prop-types';
/**
 * ErrorBoundary
 * see: https://reactjs.org/docs/error-boundaries.html
 */
export default class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    component: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null, info: null };
  }
  componentDidCatch(error, info) {
    this.setState({ hasError: true, error, info });
  }
  render() {
    if (this.state.hasError) {
      const Component = this.props.component;
      return <Component info={this.state.info} error={this.state.error} />;
    }
    return this.props.children;
  }
}
