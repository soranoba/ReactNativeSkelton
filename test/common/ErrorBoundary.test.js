import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';
import ErrorBoundary from '/common/ErrorBoundary';

describe('ErrorBoundary', () => {
  const Component = class TestComponent1 extends React.Component {
    render() {
      return <View style={{ backgroundColor: 'red' }} />;
    }
  };
  const ErrorComponent = class TestComponent2 extends React.Component {
    render() {
      return <View style={{ backgroundColor: 'green' }} {...this.props} />;
    }
  };
  const JSX = (
    <ErrorBoundary component={ErrorComponent}>
      <Component />
    </ErrorBoundary>
  );

  it('renders currectly', () => {
    const wrapper = renderer.create(JSX);
    expect(wrapper.toJSON()).toMatchSnapshot();

    wrapper.getInstance().componentDidCatch(new Error('message'), { info: 'infoMessage' });
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
