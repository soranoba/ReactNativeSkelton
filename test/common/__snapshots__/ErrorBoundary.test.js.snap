// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders currectly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "red",
    }
  }
/>
`;

exports[`ErrorBoundary renders currectly 2`] = `
<View
  error={[Error: message]}
  info={
    Object {
      "info": "infoMessage",
    }
  }
  style={
    Object {
      "backgroundColor": "green",
    }
  }
/>
`;
